### PR TITLE
Simplify Frequency Contributions

### PR DESCRIPTION
**Pull request by mgiammar**
**Adds following macros for frequency contibution**
```py
FREQ_ENUM_SHORTCUT = {
     "Shielding": {"Shielding1_0", "Shielding1_2"},
     "Isotropic": {"Shielding1_0", "J1_0"},
     "Quad": {"Quad1_2", "Quad2_0", "Quad2_2", "Quad2_4"},
     "J": {"J1_0", "J1_2"},
     "D": {"D1_2"},
     "cross": {
         "Quad_Shielding_cross_0",
         "Quad_Shielding_cross_2",
         "Quad_Shielding_cross_4",
         "Quad_J_cross_0",
         "Quad_J_cross_2",
         "Quad_J_cross_4",
         "Quad_Dipolar_cross_0",
         "Quad_Dipolar_cross_2",
         "Quad_Dipolar_cross_4",
     },
     "Quad_Shielding_cross": {
         "Quad_Shielding_cross_0",
         "Quad_Shielding_cross_2",
         "Quad_Shielding_cross_4",
     },
     "Quad_J_cross": {"Quad_J_cross_0", "Quad_J_cross_2", "Quad_J_cross_4"},
     "Quad_D_cross": {
         "Quad_Dipolar_cross_0",
         "Quad_Dipolar_cross_2",
         "Quad_Dipolar_cross_4",
     },
     "First_order": {
         "Shielding1_0",
         "Shielding1_2",
         "Quad1_2",
         "J1_0",
         "J1_2",
         "D1_2",
     },
     "Second_order": {
         "Quad2_0",
         "Quad2_2",
         "Quad2_4",
     },
     "Zeroth_rank": {
         "Shielding1_0",
         "Quad2_0",
         "J1_0",
         "Quad_Shielding_cross_0",
         "Quad_J_cross_0",
         "Quad_Dipolar_cross_0",
     },
     "Second_rank": {
         "Shielding1_2",
         "Quad1_2",
         "Quad2_2",
         "Quad_Shielding_cross_2",
         "Quad_J_cross_2",
         "Quad_Dipolar_cross_2",
     },
     "Fourth_rank": {
         "Quad2_4",
         "Quad_Shielding_cross_4",
         "Quad_J_cross_4",
         "Quad_Dipolar_cross_4",
     },
 }
```

Implements capability to group frequency contributions together and to exclude frequency contributions using a not operator `"!"` in front of the frequency contribution string

- [x] Cover new features with pytest
- [x] Update tests for freq_contrib checking (ordering of lists)
- [ ] Reach consensus on new naming conventions
- [x] Add to documentation
- [x] Update docstring

*Examples*
```
SpectralEvent(freq_contrib=["Shielding"])  # All shielding contributions
SpectralEvent(freq_contrib=["cross"])  # Only cross terms
SpectralEvent(freq_contrib=["!Quad"])  # All contributions, except quadurpolar
SpectralEvent(freq_contrib=["!cross"])  # All contributions, except cross terms
SpectralEvent(freq_contrib=["Second_rank", "!Quad"])  # All second-rank contributions, except quadurpolar
```

### CODE REVIEW

File: src/mrsimulator/method/event.py [hunk 1]
```diff
@@ -96,8 +118,14 @@ def parse_dict_with_units(cls, py_dict: dict):
...
-        array = np.zeros(len(freq_list_all), dtype=int)
-        array[[item.index() for item in self.freq_contrib]] = 1
+        """Get an array (binary) of frequency contributions flags for the event."""
+        array = np.zeros(len(FREQ_LIST_ALL), dtype=int)  # Final array to return
+
+        for item in self.freq_contrib:
+            foo = item.index()
+            array[foo] = 1
+            print(array.sum())
+        # array[[item.index() for item in self.freq_contrib]] = 1
```
Comment (deepanshs) [inline src/mrsimulator/method/event.py:128]: What was wrong with this?
```py
array[[item.index() for item in self.freq_contrib]] = 1
```

File: src/mrsimulator/method/frequency_contrib.py [hunk 2]
```diff
@@ -24,6 +24,70 @@
     "Quad_Dipolar_cross_4",
 ]
 
+FREQ_ENUM_SHORTCUT = {
+    "Shielding": {"Shielding1_0", "Shielding1_2"},
```
Comment (deepanshs) [inline src/mrsimulator/method/frequency_contrib.py:28]: Maybe also add
```"Isotropic": {"Shielding1_0", "J1_0"}```

File: src/mrsimulator/method/event.py (hunk 1 above)
Comment (mgiammar) [reply src/mrsimulator/method/event.py:128]: Oops was using that for debugging

File: src/mrsimulator/method/tests/test_event.py [hunk 3]
```diff
@@ -4,26 +4,27 @@
...
 __email__ = "srivastava.89@osu.edu"
 
 
+# TODO: Update these test to reflect sets, or ensure an ordering in freq_contrib attr
 def test_freq_contrib():
     event = BaseEvent(freq_contrib=["Quad2_4", "Quad2_0"])
     assert event.json()["freq_contrib"] == ["Quad2_4", "Quad2_0"]
     assert event.dict()["freq_contrib"] == ["Quad2_4", "Quad2_0"]
-    assert event.json(units=False)["freq_contrib"] == ["Quad2_4", "Quad2_0"]
+    assert event.json(units=False)["freq_contrib"] == ["Quad2_0", "Quad2_4"]
```
Comment (deepanshs) [inline src/mrsimulator/method/tests/test_event.py:20]: The problem with set is that it is unordered. These tests may or may not pass depending on the state of the set.

Do a set comparison instead.
```py
set(event.json(units=False)["freq_contrib"]) == set(["Quad2_0", "Quad2_4"])
```

File: src/mrsimulator/method/event.py [hunk 4]
```diff
@@ -84,6 +85,27 @@ def validate_rotor_frequency(cls, v, **kwargs):
...
+
+            # Item is a string of a recognized freq contrib string
+            elif item in FREQ_LIST_ALL:
+                additive.add(item)
+
+            # String is a Frequency Contribution shortcut
+            elif item[0] == "!":
+                subtractive.update(FREQ_ENUM_SHORTCUT[item[1:]])
+            else:
+                additive.update(FREQ_ENUM_SHORTCUT[item])
```
Comment (deepanshs) [inline src/mrsimulator/method/event.py:104]: why are you using `add` at one place and `update` at another?

**Comment by deepanshs** [comment]
The documentation for v0.8 is lagging behind. Might have to update those as a separate PR. The cross-term contribs are missing all together.

File: src/mrsimulator/method/frequency_contrib.py [hunk 5]
```diff
@@ -85,62 +150,97 @@ class FrequencyEnum(str, Enum):
...
-    #         values[item] = item
-    #     return None
-
     def json(self, **kwargs) -> str:
         """Parse the class object to a JSON compliant python dictionary object."""
         return self.value
 
+    # def dict(self, **kwargs) -> str:
+    #     """Parse the class object to a JSON compliant python dictionary object."""
+    #     return self.json()
```
Comment (deepanshs) [inline src/mrsimulator/method/frequency_contrib.py:239]: remove this

**Review by deepanshs — DISMISSED** [review]

**Comment by mgiammar** [comment]
> The documentation for v0.8 is lagging behind. Might have to update those as a separate PR. The cross-term contribs are missing all together.

I don't think the freq_contrib simplifications warrant a new simulation example, but they should be mentioned in the Method documentation. Will add to this PR just to reduce future headaches.

File: src/mrsimulator/method/frequency_contrib.py (hunk 5 above)
Comment (mgiammar) [reply src/mrsimulator/method/frequency_contrib.py]: Disregard, seralization function moved into Event class

**Review by deepanshs — APPROVED** [review]

**Pull request merged**